crates/sui-types/src/temporary_store.rs

```diff
@@ -302,16 +301,47 @@ impl<'backing> TemporaryStore<'backing> {
         }
     }
 
+    fn update_with_received_objects(
+        &mut self,
+        receiving_objects: Vec<ObjectRef>,
+        transaction_dependencies: &mut BTreeSet<TransactionDigest>,
+    ) {
+        // For any objects that we have received in this transaction, we need to add the
+        // transaction that transferred that object as a dependency.
+        let received_dependencies = receiving_objects
+            .into_iter()
+            .filter_map(|obj_ref| self.loaded_runtime_objects.get(&obj_ref.0));
+        transaction_dependencies.extend(
+            received_dependencies
+                .clone()
+                .map(|obj_meta| obj_meta.previous_transaction),
+        );
+
+        // For any received objects, we need to update the lamport timestamp for the transaction to be
+        // the maximum of the input objects and objects that have been received in the transaction.
+        // NB: this is _not_ and should not be the max of the receiving arguments of the transaction.
+        self.lamport_timestamp = self
+            .lamport_timestamp
+            .max(SequenceNumber::lamport_increment(
+                received_dependencies.map(|meta| meta.version),
+            ));
+    }
+
     pub fn to_effects(
         mut self,
         shared_object_refs: Vec<ObjectRef>,
         transaction_digest: &TransactionDigest,
-        transaction_dependencies: Vec<TransactionDigest>,
+        mut transaction_dependencies: BTreeSet<TransactionDigest>,
+        receiving_objects: Vec<ObjectRef>,
         gas_cost_summary: GasCostSummary,
         status: ExecutionStatus,
         gas_charger: &mut GasCharger,
         epoch: EpochId,
     ) -> (InnerTemporaryStore, TransactionEffects) {
+        // Update lamport version to be 1 greater than any objects that were received in the transaction.
+        // Add any received objects previous transactions as dependencies on this transaction.
+        self.update_with_received_objects(receiving_objects, &mut transaction_dependencies);
```

Comment:
@lxfind great spot on not adding dependencies! This change is what addresses your comment where we add dependencies on any transactions that transferred an object that was received during the transaction.

This also addresses another issue that I saw where we were improperly calculating the lamport timestamp for transactions -- we were not accounting for any objects that were received. This updates that logic too.

I added a number of tests to verify that we properly add dependencies, and also correctly calculate the Lamport version at the bottom of the `transfer_to_object_tests` file in the sui-core unit tests.